Framework/Geometry/src/Instrument/GridDetector.cpp

```diff
@@ -0,0 +1,939 @@
+
+  // t = coordinate along the line
+  // u,v = coordinates along horizontal, vertical
+  // (and correct for it being between 0, xpixels-1).  The +0.5 is because the
+  // base point is at the CENTER of pixel 0,0.
+  double u = (double(xpixels() - 1) * tuv[1] + 0.5);
+  double v = (double(ypixels() - 1) * tuv[2] + 0.5);
+
+  //  std::cout << u << ", " << v << "\n";
+
+  // In indices
+  int xIndex = int(u);
+  int yIndex = int(v);
+
+  // Out of range?
+  if (xIndex < 0)
+    return;
+  if (yIndex < 0)
+    return;
+  if (xIndex >= xpixels())
+    return;
+  if (yIndex >= ypixels())
+    return;
+
+  // TODO: Do I need to put something smart here for the first 3 parameters?
+  auto comp = getAtXYZ(xIndex, yIndex, 0);
+  testRay.addLink(intersec, intersec, 0.0, *(comp->shape()),
+                  comp->getComponentID());
+}
+
+//-------------------------------------------------------------------------------------------------
+//-------------------------------------------------------------------------------------------------
+// ------------ IObjComponent methods ----------------
+//-------------------------------------------------------------------------------------------------
+//-------------------------------------------------------------------------------------------------
+
+//-------------------------------------------------------------------------------------------------
+/// Does the point given lie within this object component?
+bool GridDetector::isValid(const V3D &) const {
+  throw Kernel::Exception::NotImplementedError(
+      "GridDetector::isValid() is not implemented.");
+}
+
+//-------------------------------------------------------------------------------------------------
+/// Does the point given lie on the surface of this object component?
+bool GridDetector::isOnSide(const V3D &) const {
+  throw Kernel::Exception::NotImplementedError(
+      "GridDetector::isOnSide() is not implemented.");
+}
+
+//-------------------------------------------------------------------------------------------------
+/// Checks whether the track given will pass through this Component.
+int GridDetector::interceptSurface(Track &) const {
+  throw Kernel::Exception::NotImplementedError(
+      "GridDetector::interceptSurface() is not implemented.");
+}
+
+//-------------------------------------------------------------------------------------------------
+/// Finds the approximate solid angle covered by the component when viewed from
+/// the point given
+double GridDetector::solidAngle(const V3D &) const {
+  throw Kernel::Exception::NotImplementedError(
+      "GridDetector::solidAngle() is not implemented.");
+}
+
+//-------------------------------------------------------------------------------------------------
+/// Try to find a point that lies within (or on) the object
+int GridDetector::getPointInObject(V3D &) const {
+  throw Kernel::Exception::NotImplementedError(
+      "GridDetector::getPointInObject() is not implemented.");
```

Comment:
Would be interested to hear your thoughts on these unimplemented methods.

Reply:
These methods are overridden from `IObjComponent` but are not actually used by `GridDetector` since raytracing is optimized for this (and by extension `RectangularDetector`)

Reply:
Ok so, there is some kind of capability query and then `getPointInObject` is never called.